sycl/doc/SharedLibraries.md

```diff
@@ -1,52 +1,84 @@
+use case for such linkage - providing device functions via shared libraries.
+Example:
 ```
-// App:
-
-CGH.parallel_for<app_kernel>(/* ... */ {
-  library_function();
-});
+// app.cpp
+SYCL_EXTERNAL int LibDeviceFunc(int i);
+class KernelName;
+/* ... */
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<KernelName>(/* ... */ [=](sycl::item i) {
+  out[i] = LibDeviceFunc(i);
+}); /* ... */
+std::cout << out[i] << “ “;
+
+// lib.cpp
+int SYCL_EXTERNAL LibDeviceFunc(int i) {
+  return i * 2;
+}
 
+// Commands
+clang++ -fsycl lib.cpp -shared -o helpers.so
+clang++ -fsycl app.cpp -lhelpers -o a.out
+./a.out
+Output: 0 2 4 6…
 
-// Shared library:
-SYCL_EXTERNAL void library_function() {
-  // do something
-}
 ```
 It is possible to manually create `sycl::program` in both app and shared
 library, then use `link` SYCL API to get a single program and launch kernels
 using it. But it is not user-friendly and it is very different from regular
 C/C++ workflow.
 
+Another possible scenario - use functions defined in pre-compiled device image
+provided by user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void foo();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { foo(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void foo() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option (like `-fsycl-add-targets`).
+
 The main purpose of this feature is to provide a mechanism which allows to
-provide *device* functions via shared libraries and works as close as possible
-to regular shared libraries.
-
-## Requrements:
-User's code is compiled into a shared library which consists of some host API,
-device code and device API (`SYCL_EXTERNAL` functions). The library is linked to
-a user's application which also contains some device code and performs
-computations using DPC++/SYCL.
+link device code dynamically at runtime.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
+dynamically at run time with device code of a user's application in order to
+resolve dependencies.
 For this combination the following statements must be true:
 
-- `SYCL_EXTERNAL` functions from library can be called (directly or indirectly)
-  from device code of the application.
-- Function pointers taken in application should work inside the library.
+- `SYCL_EXTERNAL` functions defined in dynamically linked code can be called
+  (directly or indirectly) from device code of the application.
+- Function pointers taken in application should work inside the dynamically
+  linked code.
```

Comment:
I'm not sure about this point:
> (NOTE: I think this one would be really useful, I suggest to add a TODO as potential design extension point) loaded into memory via analog of dlopen

Do you mean some specific SYCL API? Because if dlopen is used it will be just specific usage of dynamically linked code that "embedded into a host shared object by standard SYCL compiler driver invocation".

Reply:
I meant some new API, which would load a device library from disk (optionally returning `device_image`), and automatically make symbols defined in the loaded binary image available to SYCL runtime for resolution:
```cpp
// suppose, mylib.spv defines SYCL_EXTERNAL function foo, then this call:
device_image img = device_dlopen("mylib.spv");
// will make foo available for dynamic symbol resolution. If any subsequent JIT compilations
// try to compile device code with external reference to foo, it can now be resolved following
// the resolution mechanism described in this doc, and JIT compilation will succeed.
```

Reply:
Ok, thanks. I mentioned that.

Reply:
Do we really need a new API? Can't we just suggest compiling marked with SYCL_EXTERNAL as a regular "fat" .so and then dlopen it? dlopen'ing should trigger device image registration, so such a device image should be available for symbols resolution.

Reply:
> Can't we just suggest compiling marked with SYCL_EXTERNAL as a regular "fat" .so and then dlopen it?

Sorry for delay. Yes, you are right, that should be the preferred way. But sometimes there is a need to load .spv originating from different tools - e.g. OpenCL. It would be UB trying to define kernels in this .spv and then trying to call them from SYCL, but it should be perfectly OK to link them with .spv originating from SYCL.